calblueprint/immigration-justice-project
# [feat] create user onboarding context

[//]: # "These comments are meant for your reference. They are invisible and don't need to be deleted!"

## :tanabata_tree: Description

### :palm_tree: What's new in this PR
[//]: # "REQUIRED - Describe what's new in this PR in a few lines. A description and bullet points for specifics will suffice."

Created `src/utils/OnboardingProvider.tsx`:
- Added context and provider
- Added example usage

Modified `src/app/onboarding/layout.tsx`:
- Use `OnboardingProvider.tsx`
- Offload work to `OnboardingManager` component (which can use onboarding context)

Created `src/components/OnboardingManager.tsx`:
- Handles routing changes and progress bar updates
- Protection to prevent modifying URL to advance (not using buttons)
    - Currently allows backtracking with back button

Modified `src/api/supabase/queries/profiles.ts` to support relevant CRUD operations.

Replaced default Vercel icon with IJP favicon.

### :evergreen_tree: Screenshots
[//]: # "REQUIRED for frontend changes, otherwise optional but strongly recommended. Add screenshots of expected behavior - GIFs if you're feeling fancy! If you are making changes here, please CC: @davidqing6432 at the bottom."

![demo_onboarding](https://github.com/calblueprint/immigration-justice-project/assets/67077248/955bd9b7-89d3-4621-8ed2-499985581a59)

![image](https://github.com/calblueprint/immigration-justice-project/assets/67077248/5b497c94-a7df-41ec-be77-a50f0d1979e3)


## :deciduous_tree: How to review
[//]: # 'REQUIRED - Describe the order in which to review files and what to expect when testing locally. Is there anything specifically you want feedback on? Should this be reviewed commit by commit, or all at once? What are some user flows to test? What are some edge cases to look out for?'

#### Code Review

First review changes in queries, then review provider, then `OnboardingManager`.

#### Testing

Standard procedure:
```
git fetch origin jinkang/ijp-73-create-user-onboarding-context
git checkout jinkang/ijp-73-create-user-onboarding-context
```

#### Testing Context

Testing setup:
- Setup according to example usage at the bottom of `src/utils/OnboardingProvider.tsx`
    - Remember to setup both `layout.tsx` and `page.tsx`
- Make sure you are signed in to an account (navigate to `/login` to do so)
- Make sure you do NOT have a row associated with your user ID in the profiles table

When all of the above are complete, start the app with `npm run dev` and navigate to the testing page to try it out.

#### Testing Routing

Navigate to any of the following and test it out:
- `/onboarding/roles`
- `/onboarding/basic-info`
- `/onboarding/availability`
- `/onboarding/legal-experience`
- `/onboarding/done`

You should notice that you're redirected to roles regardless of which route you start at.

Try to break it to the greatest extent that you can.


## :seedling: Next steps 
[//]: # "Optional - What's NOT in this PR, doesn't work yet, and/or still needs to be done. Note any temporary fixes in this PR that should be cleaned up later."

Implement dynamically updating flow with Roles page.


## :link: Relevant Links



### :information_source: Online sources
[//]: # 'Optional - copy links to any tutorials or documentation that was useful to you when working on this PR'



### :potted_plant: Related PRs
[//]: # "Optional - related PRs you're waiting on/ PRs that will conflict, etc; if this is a refactor, feel free to add PRs that previously modified this code"



[//]: # 'This tags the project leader as a default. Feel free to change, or add on anyone who you should be in on the conversation.'
CC: @pragyakallanagoudar
